### ISSUE TITLE
resource reporting vs docker cookbook

### ISSUE THREAD

**Issue by b-dean** [sethvargo/chefspec#676]
The docker cookbook version  2.5.7 (and 2.5.8) introduced code in `libaries/_autoload.rb` that uses resources to install some gems.
* https://github.com/chef-cookbooks/docker/blob/v2.5.7/libraries/_autoload.rb
* https://github.com/chef-cookbooks/docker/blob/v2.5.8/libraries/_autoload.rb

It seems like the resource reporting of chefspec has problems with this because there isn't a node.runner at that point in the cookbook compile.

```
     NoMethodError:
       undefined method `step_into?' for nil:NilClass
     # ./vendor/bundle/ruby/2.3.0/gems/chefspec-4.5.0/lib/chefspec/extensions/chef/resource.rb:20:in `run_action'
     # /tmp/d20160229-5221-14o7666/cookbooks/docker/libraries/_autoload.rb:10:in `rescue in <top (required)>'
     # /tmp/d20160229-5221-14o7666/cookbooks/docker/libraries/_autoload.rb:1:in `<top (required)>'
     # ./vendor/bundle/ruby/2.3.0/gems/chef-12.7.2/lib/chef/run_context/cookbook_compiler.rb:191:in `load'
     # ./vendor/bundle/ruby/2.3.0/gems/chef-12.7.2/lib/chef/run_context/cookbook_compiler.rb:191:in `block in load_libraries_from_cookbook'
     # ./vendor/bundle/ruby/2.3.0/gems/chef-12.7.2/lib/chef/run_context/cookbook_compiler.rb:188:in `each'
     # ./vendor/bundle/ruby/2.3.0/gems/chef-12.7.2/lib/chef/run_context/cookbook_compiler.rb:188:in `load_libraries_from_cookbook'
     # ./vendor/bundle/ruby/2.3.0/gems/chef-12.7.2/lib/chef/run_context/cookbook_compiler.rb:99:in `block in compile_libraries'
     # ./vendor/bundle/ruby/2.3.0/gems/chef-12.7.2/lib/chef/run_context/cookbook_compiler.rb:98:in `each'
     # ./vendor/bundle/ruby/2.3.0/gems/chef-12.7.2/lib/chef/run_context/cookbook_compiler.rb:98:in `compile_libraries'
     # ./vendor/bundle/ruby/2.3.0/gems/chef-12.7.2/lib/chef/run_context/cookbook_compiler.rb:71:in `compile'
     # ./vendor/bundle/ruby/2.3.0/gems/chef-12.7.2/lib/chef/run_context.rb:167:in `load'
     # ./vendor/bundle/ruby/2.3.0/gems/chef-12.7.2/lib/chef/policy_builder/expand_node_object.rb:93:in `setup_run_context'
     # ./vendor/bundle/ruby/2.3.0/gems/chef-12.7.2/lib/chef/client.rb:509:in `setup_run_context'
     # ./vendor/bundle/ruby/2.3.0/gems/chefspec-4.5.0/lib/chefspec/solo_runner.rb:117:in `converge'
     # ./vendor/bundle/ruby/2.3.0/gems/chefspec-4.5.0/lib/chefspec/solo_runner.rb:18:in `block in converge'
     # ./vendor/bundle/ruby/2.3.0/gems/chefspec-4.5.0/lib/chefspec/solo_runner.rb:17:in `tap'
     # ./vendor/bundle/ruby/2.3.0/gems/chefspec-4.5.0/lib/chefspec/solo_runner.rb:17:in `converge'
     # ./spec/foo_spec.rb:5:in `block (2 levels) in <top (required)>'
     # ./spec/foo_spec.rb:8:in `block (2 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # Gem::LoadError:
     #   docker-api is not part of the bundle. Add it to Gemfile.
     #   /home/ben/.rvm/gems/ruby-2.3.0@global/gems/bundler-1.10.6/lib/bundler/rubygems_integration.rb:292:in `block in replace_gem'
```

Here's a gist with the files I used to make this happen: https://gist.github.com/b-dean/237e97ae01e8d8e4f45e

**Comment by b-dean** [comment]
I should also point out that a workaround is to just put `docker-api` in your Gemfile so the docker cookbook finds it while it's compiling

```ruby
gem 'docker-api'
```

**Comment by sethvargo** [comment]
Oh yea... Creating a new RunContext is totally going to break ChefSpec. Unfortunately I don't know of a good way around that. ChefSpec needs a solid run context in order to introspect the Chef Client run.

**Comment by b-dean** [comment]
Maybe `Chef::RunContext.new` should be monkey patched so ChefSpec knows about any run context that is created?

**Comment by sethvargo** [comment]
Hi @b-dean 

I'm sorry, but I refuse to monkey patch a monkey patch because someone else is already monkey patching...

**Comment by b-dean** [comment]
They aren't really monkey patching anything. Are you saying that cookbooks make their own `Chef::RunContext` objects (which seems like a thing that can be done with Chef) are incompatible w/ ChefSpec?

**Comment by chasebolt** [comment]
The docker Gemfile depends on the docker-api, which chefspec will then load and the error will go away.

tl;dr run bundle install

**Comment by b-dean** [comment]
I did a `bundle install`, but my Gemfile doesn't have `docker-api`, see [the gist I linked to in the issue](https://gist.github.com/b-dean/237e97ae01e8d8e4f45e).

**Comment by sethvargo** [comment]
Looks like this was fixed upstream
